engine/consensus/dkg/dkg_controller_test.go

```diff
@@ -0,0 +1,273 @@
+		testCase{totalNodes: 15, phase0Duration: 5 * time.Second, phase1Duration: 2 * time.Second, phase2Duration: 2 * time.Second},
+	}
+
+	// run each test case
+	for _, tc := range testCases {
+		// gn is the minimum number of good nodes required for the DKG protocol
+		// to go well
+		gn := tc.totalNodes - optimalThreshold(tc.totalNodes)
+
+		t.Run(fmt.Sprintf("%d/%d nodes", gn, tc.totalNodes), func(t *testing.T) {
+			testDKG(t, tc.totalNodes, gn, tc.phase0Duration, tc.phase1Duration, tc.phase2Duration)
+		})
+	}
+}
+
+func testDKG(t *testing.T, totalNodes int, goodNodes int, phase0Duration, phase1Duration, phase2Duration time.Duration) {
+	nodes := initNodes(t, totalNodes, phase0Duration, phase1Duration, phase2Duration)
+	gnodes := nodes[:goodNodes]
+
+	// Start all the good nodes in parallel
+	for _, n := range gnodes {
+		go func(node *node) {
+			err := node.run()
+			require.NoError(t, err)
+		}(n)
+	}
+
+	// Wait until they are all shutdown
+	wait(t, gnodes, 5*phase0Duration)
+
+	// Check that all nodes have agreed on the same set of public keys
+	checkArtifacts(t, gnodes, totalNodes)
+}
+
+// Initialise nodes and communication channels.
+func initNodes(t *testing.T, n int, phase0Duration, phase1Duration, phase2Duration time.Duration) []*node {
+	// Create the channels through which the nodes will communicate
+	channels := make([]chan DKGMessage, 0, n)
+	for i := 0; i < n; i++ {
+		channels = append(channels, make(chan DKGMessage, 5*n*n))
+	}
+
+	nodes := make([]*node, 0, n)
+
+	// Setup
+	for i := 0; i < n; i++ {
+		seed := make([]byte, 20)
+		_, _ = rand.Read(seed)
+
+		processor := &processor{
+			id:       i,
+			channels: channels,
+		}
+
+		dkg, err := crypto.NewJointFeldman(n, optimalThreshold(n), i, processor)
+		require.NoError(t, err)
+
+		controller := NewController(
+			dkg,
+			seed,
+			channels[i],
+			zerolog.New(os.Stderr).With().Int("id", i).Logger())
+
+		node := newNode(i, controller, phase0Duration, phase1Duration, phase2Duration)
+
+		nodes = append(nodes, node)
+	}
+
+	return nodes
+}
+
+// Wait for all the nodes to reach the SHUTDOWN state, or timeout.
+func wait(t *testing.T, nodes []*node, timeout time.Duration) {
+
+	timer := time.After(timeout)
+
+	for {
+		select {
+		case <-timer:
+			t.Fatal("TIMEOUT")
+		default:
+			done := true
+			for _, node := range nodes {
+				if node.controller.GetState() != Shutdown {
+					done = false
+					break
+				}
+			}
+			if done {
+				return
+			}
+			time.Sleep(1 * time.Second)
+		}
+	}
+}
+
+// Check that all nodes have produced the same set of public keys
+func checkArtifacts(t *testing.T, nodes []*node, totalNodes int) {
+	for i := 1; i < len(nodes); i++ {
+		require.NotEmpty(t, nodes[i].controller.privateShare)
+		require.NotEmpty(t, nodes[i].controller.groupPublicKey)
+
+		require.Equal(t, nodes[0].controller.groupPublicKey, nodes[i].controller.groupPublicKey,
```

Comment:
sorry my earlier comment wasn't clear, I meant this: 
( it's actually the correct test for keys, `require.Equal` or `DeepEqual` may fail although keys are formally equal )
```suggestion
		require.True(t, nodes[0].controller.groupPublicKey.Equals( nodes[i].controller.groupPublicKey),
```

Reply:
and the same applies for line 255